Docs[Pomegranate]: Add infra info (#4)
paastech-cloud/docs: Architecture.md

#### Client applications deployment

The application manager, Pomegranate is charged with acting on the containers' lifecycle.

[Rust](https://www.rust-lang.org/) was used to develop the service. We used this programming language because
we wanted a performant language that could be trusted with this job thanks to its compiler,

> and with support for the other technologies we use.

Author: While I appreciate the stylistic changes, one important bit of information is dropped here: how strong the compiler constraints are.
 I'll accept it nonetheless edit it afterward if I find a good way to articulate my point

> we wanted an efficient and reliable language that could be trusted with this job thanks to its compiled nature and the restrictions it brings to drive us to write better code,

Reviewer: This sentence is too long and complex. The notion is good, but it has to be broken down and explained better. Also, *how* exactly do the restrictions help to write better code?

Author: Explaining all the things the compiler does to ~~torture~~ care after us would be a long task.
Perhaps I should link to an article that dives into the details ?

Reviewer: Explaining a bit more some details could be great, and I do approve of the idea of inserting a link to a detailed article on this topic. Please ensure the source is trustworthy beforehand.

Author: For some reason I couldn't find a good article at all

Author: I broke down the sentences but didn't really try to elaborate a lot on the compiler. Nonetheless a new review is needed

> The [Rust](https://www.rust-lang.org/) programming language is used to develop the service. We used it because
> we wanted an efficient language that could be trusted with this job thanks to its compiled nature,
> and with support for the other technologies we use through the crates ecosystem.

Reviewer 2: What do you think about this for talking about compiler constraints?
```suggestion
The [Rust](https://www.rust-lang.org/) programming language is used to develop the service. We used it because
we wanted an efficient and reliable language that could be trusted with this job thanks to its compiled nature and the restrictions it brings to drive us to write better code,
and with support for the other technologies we use through the crates ecosystem.
```

> and with support for the other technologies we use through the crates ecosystem.

Reviewer: Define the crates ecosystem, explain what are crates and what Cargo is

Author: ![image](https://github.com/paastech-cloud/docs/assets/24864674/00cb581e-159d-444f-b795-4762d234a7ed)
Is that enough ?

Author: Should be done

Establishing communication with other services is done via [gRPC](https://grpc.io/docs/what-is-grpc/core-concepts/).

> gRPC imposed itself as the go-to framework for internal APIs thanks to its [high performance](https://www.nexthink.com/blog/comparing-grpc-performance)

Reviewer: Can you please present the benchmarks? Having a link is good, but a sum up could help the reader understand our point. Also, the sentence is missing a full stop.

Author: I have to say, it feels wasteful in my opinion. Perhaps I should just include the performance graph with a small caption ?

Reviewer: Including the performance graph seems quite a bit of work to add, since the image needs to be placed in a separate folder in the codebase. I was thinking that maybe a quick overview of the benchmark could give a first opinion on its truthfulness, while still giving some support to your claims.

> To handle containerized client applications. We chose [Docker](https://www.docker.com/) because we already

Reviewer: Why add the full stop at the middle of the sentence? This should be a comma

Author: Indeed, this is a typo

knew how to use it, and it is straightforward to communicate with the Docker server to manage containers.
It additionally provides us with easy networking and storage APIs, so we don't have to worry about those issues.

> This choice helped us to quickly set up the MVP and get a working product as soon as possible.

Reviewer: ```suggestion
This choice helped us to quickly set up the MVP (Minimum Viable Product) and get a working product as soon as possible.
```

> Communication to the docker socket was achieved through the Rust crate [Bollard](https://crates.io/crates/bollard/).

Reviewer: ```suggestion
Communication to the Docker socket was achieved through the [Bollard](https://crates.io/crates/bollard/) Rust crate.
```

#### Client applications deployment

Pomegranate, the application and deployment manager, is in charge of monitoring and managing the containers' lifecycle, as well as configuring them.

The [Rust](https://www.rust-lang.org/) programming language is used to develop the service.
We used it because we wanted an efficient and reliable language that could be trusted.

> Thanks to its strict compiler, we are forced to write better code that handle errors and we get extreme memory safety.

Reviewer 2: The wording is too negative imho. Also, rephrased a bit.

```suggestion
Thanks to its compiler, we are driven to write better code that handles errors and memory safety is guaranteed.
```

Furthermore, it supports the other technologies we use through the crates ecosystem.

> Crates are third-party packages providing additional functionalities and are managed by Cargo to track crates and their dependencies. 

Reviewer 2: ```suggestion
Crates are third-party packages that provides additional features. They are managed by Cargo, which handles dependency resolution and versioning.
```

Establishing communication with other services is done via [gRPC](https://grpc.io/docs/what-is-grpc/core-concepts/).
gRPC imposed itself as the go-to framework for internal APIs thanks to its [high performance](https://www.nexthink.com/blog/comparing-grpc-performance) compared to [HTTP performance](https://github.com/programatik29/rust-web-benchmarks/blob/master/result/hello-world.md#comparisons).

> The synthetic benchmarks show:

Reviewer 2: ```suggestion
Synthetic benchmarks shows that:
```

>  - gRPC performs ~25% more requests compared to HTTP on a single core

Reviewer 2: ```suggestion
 - gRPC processes ~25% more requests compared to HTTP on a single core
```

#### Client Applications

Pomegranate is responsible for starting, stopping, and interacting with the client applications.
It is also responsible for managing the networking of the applications.

Pomegranate is completely stateless, and the API is the only external way to interact with it.
Once a container is started, Pomegranate does not keep track of it, and it is the responsibility of the execution engine to manage it.

```mermaid
flowchart TD

>     web[80]
>     webscure[443]

Reviewer: Add named "ports" to specify to the reader those are machine ports

```suggestion
    web([port 80])
    webscure([port 443])
```

    traefik[Traefik]
    app_1[User app 1]
    app_2[User app 2]
    app_3[User app 3]

    rules{Resolve\nvia labels}

    webscure --> |HTTPS| traefik
    web --> |HTTP| traefik
    traefik --> rules

>     rules --> |uuid_1.user-app:80| app_1
>     rules --> |uuid_2.user-app:80| app_2
>     rules --> |uuid_3.user-app:80| app_3

Reviewer: Suggest adding bold to UUIDs to make them stand out, and add the FQDN to the end of the URL

```suggestion
    rules --> |"`**uuid1**.user-app.**FQDN**:80`"| app_1
    rules --> |"`**uuid2**.user-app.**FQDN**:80`"| app_2
    rules --> |"`**uuid3**.user-app.**FQDN**:80`"| app_3
```

Author: While the suggestion makes sense for the uuid, the FQDN part is just overloading the graph for very little added value.

Reviewer: We can find a compromise in that, however I would still suggest defining FQDN elsewhere (maybe after the graph) in order for the reader to follow along.

```

To expose newly spawned containers to the World Wide Web, [Traefik](https://doc.traefik.io/traefik/) is configured to answer on 80 and 443 ports.

> Then a set of labels is attributed to each container to create a unique subdomain `<app_uuid>.user-app.<fqdn>` redirecting to the port 80 of the associated app.

Reviewer: Define who handles or transmits the FQDN to Pomegranate. If not given to Pomegranate, where should it fit and what is it?

Author: It is given to Pomegranate through an env variable, is this really relevant to state ?
The reason why `<fqdn>` was used is for agnosticism of the environment where it is deployed.

Reviewer: I agree with the agnosticism, but the FQDN lacks a proper definition and leaves the reader scratching their head. Could this be improved with a quick run down of how agnosticism has been applied to Promegranate, and how the FQDN allows it to be versatile environment-wise?

Author: Should be done

#### Client Applications

Pomegranate is responsible for starting, stopping, and interacting with the client applications.
It is also responsible for managing the networking of the applications.

Pomegranate is completely stateless, and the API is the only external way to interact with it.
Once a container is started, Pomegranate does not keep track of it, and it is the responsibility of the execution engine to manage it.

```mermaid
flowchart TD
    web([port 80])
    webscure([port 443])
    traefik[Traefik]
    app_1[User app 1]
    app_2[User app 2]
    app_3[User app 3]

    rules{Resolve\nvia labels}

    webscure --> |HTTPS| traefik
    web --> |HTTP| traefik
    traefik --> rules

    rules --> |**uuid_1**.user-app:80| app_1
    rules --> |**uuid_2**.user-app:80| app_2
    rules --> |**uuid_3**.user-app:80| app_3

```

> The [FQDN (Fully Qualified Domain Name)](https://en.wikipedia.org/wiki/Fully_qualified_domain_name) is passed to Pomegranate via an environment variable.

Reviewer 2: Avoid directly saying "FQDN", prefer "domain name". The environment variable is not even an FQDN (the domain name you pass doesn't end with a `.` character).

Also, by reading this, I don't know what domain name you refer to nor how it is being used.

Author: rip

Doing so enables flexibility during the development process, as the developer can easily swap between a local and a production environment.

> To expose newly spawned containers to the World Wide Web, [Traefik](https://doc.traefik.io/traefik/) is configured to answer on 80 and 443 ports.

Reviewer 2: ```suggestion
To expose newly spawned containers to the World Wide Web, [Traefik](https://doc.traefik.io/traefik/) is configured to listen to the 80 and 443 ports.
```

Then a set of labels is attributed to each container to create a unique subdomain `<app_uuid>.user-app.<fqdn>` redirecting to the port 80 of the associated app.

> TLS termination is handled by Traefik, by resolving the DNS-01 challenge with the TLS provider of choice. This allows us to have a valid certificate for all subdomains of PaaSTech.

Reviewer 2: ```suggestion
TLS termination is handled by Traefik. It manages resolving the DNS-01 challenge with the configured certificate authority. This allows us to have a valid certificate for all subdomains of PaaSTech.
```

> In this case, the DNS registrar is Porkbun.

Reviewer 2: ```suggestion
In this case, our DNS registrar is Porkbun, and our certificate authority is Let's Encrypt.
```

#### Client Applications

> Pomegranate, our application manager, is responsible for starting, stopping, and interacting with the client applications.

Reviewer 2: ```suggestion
Pomegranate is responsible for starting, stopping, and interacting with the client applications.
```

It is also responsible for managing the networking of the applications.

Pomegranate is completely stateless, and the API is the only external way to interact with it.
Once a container is started, Pomegranate does not keep track of it, and it is the responsibility of the execution engine to manage it.

```mermaid
flowchart TD
    web[80]
    webscure[443]
    traefik[Traefik]
    app_1[User app 1]
    app_2[User app 2]
    app_3[User app 3]

    rules{Resolve\nvia labels}

    webscure --> |HTTPS| traefik
    web --> |HTTP| traefik
    traefik --> rules

    rules --> |uuid_1.user-app:80| app_1
    rules --> |uuid_2.user-app:80| app_2
    rules --> |uuid_3.user-app:80| app_3

```

To expose newly spawned containers to the World Wide Web, [Traefik](https://doc.traefik.io/traefik/) is configured to answer on 80 and 443 ports.
Then a set of labels is attributed to each container to create a unique subdomain `<app_uuid>.user-app.<fqdn>` redirecting to the port 80 of the associated app.

> TLS Termination is handled by Traefik, by resolving the DNS-O1 challenge with the TLS provider of choice. This allows us to have a valid certificate for all subdomains of PaasTech.

Reviewer 2: ```suggestion
TLS Termination is handled by Traefik, by resolving the DNS-O1 challenge with the TLS provider of choice. This allows us to have a valid certificate for all subdomains of PaaSTech.
```

> In this case, the provider is Porkbun.

Reviewer 2: ```suggestion
In this case, the DNS registrar is Porkbun.
```

## Post-mortem

### Organisational overview

***TODO: [UNIFIED WORK] from an organisation standpoint, how was the entire team organised, how did the squads interact***

### Infrastructure

**Lessons learned:**

- Perhaps Rust was not the best language to interact with Docker since docker/kube has a first-party API in Golang. We had a lot of Rust specific issues during development that slowed down how productive we were.
- We defined poorly how our service should be interacted with, forcing us to adapt interfaces at the last minute.

> - Docker is far from ideal to build a PaaS on. A more reliable and secure option would be to use Kubernetes, ideally with MicroVMs.

Reviewer 2: ```suggestion
- Docker is far from ideal to build a PaaS on. A more reliable and secure option would be to use Kubernetes, ideally with microVMs.
```
